File: lib/rekindle/toolchain/exec.ex
defmodule Rekindle.Toolchain.Exec do
  @moduledoc false

  alias Rekindle.Toolchain.Frame

  @cancel_reasons ~w[obsolete timeout shutdown caller]
  @env_name ~r/\A[A-Za-z_][A-Za-z0-9_]{0,127}\z/

  defstruct request_id: nil,
            phase: :awaiting_started,
            output_bytes_per_stream: 16_777_216,
            pid: nil,
            process_group: nil,
            stdout_sequence: 0,
            stderr_sequence: 0,
            stdout_eof?: false,
            stderr_eof?: false,
            stdout_bytes: 0,
            stderr_bytes: 0,
            terminal: nil

  @type t :: %__MODULE__{}

  @spec spawn_request(keyword()) :: {:ok, map(), t()} | {:error, atom()}
  def spawn_request(options) do
    request_id = Keyword.get_lazy(options, :request_id, &random_id/0)
    executable = Keyword.get(options, :executable)
    argv = Keyword.get(options, :argv, [])
    cwd = Keyword.get(options, :cwd)
    env_mode = Keyword.get(options, :env_mode, :replace)
    env_set = Keyword.get(options, :env_set, [])
    env_unset = Keyword.get(options, :env_unset, [])
    terminate_grace = Keyword.get(options, :terminate_grace_ms, 3_000)
    kill_grace = Keyword.get(options, :kill_grace_ms, 2_000)
    output_bytes = Keyword.get(options, :output_bytes_per_stream, 16_777_216)

    with :ok <- request_id(request_id),
         :ok <- absolute(executable),
         :ok <- absolute(cwd),
         :ok <- strings(argv),
         true <- env_mode in [:inherit, :replace],
         {:ok, env_set} <- env_set(env_set),
         {:ok, env_unset} <- env_unset(env_unset),
         true <- MapSet.disjoint?(MapSet.new(env_set, &elem(&1, 0)), MapSet.new(env_unset)),
         true <- integer_between?(terminate_grace, 0, 30_000),
         true <- integer_between?(kill_grace, 100, 30_000),
         true <- integer_between?(output_bytes, 1_048_576, 268_435_456) do
      header = %{
        "v" => 1,
        "type" => "spawn",
        "request_id" => request_id,
        "payload_len" => 0,
        "executable" => %{"kind" => "path", "value" => executable},
        "argv" => argv,
        "cwd" => cwd,
        "env_mode" => Atom.to_string(env_mode),
        "env_set" => Enum.map(env_set, fn {key, value} -> [key, value] end),
        "env_unset" => env_unset,
        "terminate_grace_ms" => terminate_grace,
        "kill_grace_ms" => kill_grace
      }

      {:ok, header, %__MODULE__{request_id: request_id, output_bytes_per_stream: output_bytes}}
    else
      _ -> {:error, :invalid_spawn}
    end
  end

  @spec cancel(t(), atom()) :: {:ok, map()} | {:error, atom()}
  def cancel(%__MODULE__{terminal: nil, request_id: request_id}, reason)
      when is_atom(reason) do
    reason = Atom.to_string(reason)

    if reason in @cancel_reasons do
      {:ok,
       %{
         "v" => 1,
         "type" => "cancel",
         "request_id" => request_id,
         "payload_len" => 0,
         "reason" => reason
       }}
    else
      {:error, :invalid_cancel}
    end
  end

  def cancel(_state, _reason), do: {:error, :terminal_session}

  @spec accept(t(), map(), binary()) :: {:ok, t()} | {:terminal, map(), t()} | {:error, atom()}
  def accept(%__MODULE__{terminal: terminal}, _header, _payload) when not is_nil(terminal),
    do: {:error, :post_terminal_frame}

  def accept(state, header, payload) when is_map(header) and is_binary(payload) do
    with :ok <- same_request(state, header),
         true <- header["payload_len"] == byte_size(payload) do
      accept_type(state, header, payload)
    else
      _ -> {:error, :invalid_exec_frame}
    end
  end

  defp accept_type(
         %__MODULE__{phase: :awaiting_started} = state,
         %{"type" => "started"} = header,
         <<>>
       ) do
    if exact_keys?(header, ~w[v type request_id payload_len pid process_group]) and
         positive(header["pid"]) and positive(header["process_group"]) do
      {:ok,
       %{state | phase: :streaming, pid: header["pid"], process_group: header["process_group"]}}
    else
      {:error, :invalid_started}
    end
  end

  defp accept_type(
         %__MODULE__{phase: :awaiting_started} = state,
         %{"type" => "exit"} = header,
         <<>>
       ) do
    terminal(state, header, spawn_failed?: true)
  end

  defp accept_type(
         %__MODULE__{phase: :streaming} = state,
         %{"type" => type} = header,
         payload
       )
       when type in ["stdout", "stderr"] do
    stream(state, header, payload)
  end

  defp accept_type(
         %__MODULE__{phase: :streaming} = state,
         %{"type" => "exit"} = header,
         <<>>
       ) do
    terminal(state, header, spawn_failed?: false)
  end

  defp accept_type(_state, _header, _payload), do: {:error, :unexpected_exec_frame}

  defp stream(state, header, payload) do
    stream = header["type"]
    sequence_key = if stream == "stdout", do: :stdout_sequence, else: :stderr_sequence
    eof_key = if stream == "stdout", do: :stdout_eof?, else: :stderr_eof?
    bytes_key = if stream == "stdout", do: :stdout_bytes, else: :stderr_bytes

    cond do
      not exact_keys?(header, ~w[v type request_id payload_len sequence eof]) ->
        {:error, :invalid_stream_frame}

      Map.fetch!(state, eof_key) ->
        {:error, :stream_after_eof}

      header["sequence"] != Map.fetch!(state, sequence_key) ->
        {:error, :stream_sequence}

      not is_boolean(header["eof"]) ->
        {:error, :invalid_stream_frame}

      true ->
        {:ok,
         state
         |> Map.put(sequence_key, Map.fetch!(state, sequence_key) + 1)
         |> Map.put(eof_key, header["eof"])
         |> Map.put(bytes_key, Map.fetch!(state, bytes_key) + byte_size(payload))}
    end
  end

  defp terminal(state, header, spawn_failed?: spawn_failed?) do
    required =
      ~w[v type request_id payload_len outcome code signal cleanup stdout_bytes stderr_bytes discarded_stdout discarded_stderr]

    outcome = header["outcome"]

    cond do
      not exact_keys?(header, required) ->
        {:error, :invalid_exit}

      outcome not in ~w[exited signaled spawn_failed] ->
        {:error, :invalid_exit}

      spawn_failed? != (outcome == "spawn_failed") ->
        {:error, :invalid_exit_order}

      not spawn_failed? and not (state.stdout_eof? and state.stderr_eof?) ->
        {:error, :missing_stream_eof}

      outcome == "exited" and (not is_integer(header["code"]) or not is_nil(header["signal"])) ->
        {:error, :invalid_exit}

      outcome == "signaled" and (not is_integer(header["signal"]) or not is_nil(header["code"])) ->
        {:error, :invalid_exit}

      outcome == "spawn_failed" and (not is_nil(header["code"]) or not is_nil(header["signal"])) ->
        {:error, :invalid_exit}

      header["cleanup"] not in ~w[confirmed uncertain] ->
        {:error, :invalid_exit}

      header["stdout_bytes"] != state.stdout_bytes or header["stderr_bytes"] != state.stderr_bytes ->
        {:error, :byte_count_mismatch}

      not nonnegative(header["discarded_stdout"]) or not nonnegative(header["discarded_stderr"]) ->
        {:error, :invalid_exit}

      true ->
        result = %{
          outcome: decode_outcome(outcome),
          code: header["code"],
          signal: header["signal"],
          cleanup: decode_cleanup(header["cleanup"]),
          stdout_bytes: header["stdout_bytes"],
          stderr_bytes: header["stderr_bytes"],
          discarded_stdout: header["discarded_stdout"],
          discarded_stderr: header["discarded_stderr"]
        }

        {:terminal, result, %{state | phase: :terminal, terminal: result}}
    end
  end

  defp decode_outcome("exited"), do: :exited
  defp decode_outcome("signaled"), do: :signaled
  defp decode_outcome("spawn_failed"), do: :spawn_failed

  defp decode_cleanup("confirmed"), do: :confirmed
  defp decode_cleanup("uncertain"), do: :uncertain

  @spec encode(map()) :: {:ok, binary()} | {:error, atom()}
  def encode(header), do: Frame.encode(header)

  defp same_request(state, header) do
    if header["request_id"] == state.request_id and header["v"] == 1,
      do: :ok,
      else: {:error, :request_mismatch}
  end

  defp request_id(value) when is_binary(value) do
    if Regex.match?(~r/\A[0-9a-f]{32}\z/, value), do: :ok, else: {:error, :request_id}
  end

  defp request_id(_), do: {:error, :request_id}

  defp absolute(value) when is_binary(value) do
    if Path.type(value) == :absolute and String.valid?(value) and
         not String.contains?(value, <<0>>), do: :ok, else: {:error, :path}
  end

  defp absolute(_), do: {:error, :path}

  defp strings(values) when is_list(values) do
    if proper_list?(values) and
         Enum.all?(
           values,
           &(is_binary(&1) and String.valid?(&1) and not String.contains?(&1, <<0>>))
         ), do: :ok, else: {:error, :strings}
  end

  defp strings(_), do: {:error, :strings}

  defp env_set(values) when is_list(values) do
    with true <- proper_list?(values),
         true <-
           Enum.all?(values, fn
             {key, value} ->
               valid_env?(key) and is_binary(value) and String.valid?(value) and
                 not String.contains?(value, <<0>>)

             _malformed ->
               false
           end),
         true <- unique?(Enum.map(values, &elem(&1, 0))) do
      {:ok, Enum.sort_by(values, &elem(&1, 0))}
    else
      _ -> {:error, :environment}
    end
  end

  defp env_set(_), do: {:error, :environment}

  defp env_unset(values) when is_list(values) do
    if proper_list?(values) and Enum.all?(values, &valid_env?/1) and unique?(values),
      do: {:ok, Enum.sort(values)},
      else: {:error, :environment}
  end

  defp env_unset(_), do: {:error, :environment}

  defp valid_env?(value), do: is_binary(value) and Regex.match?(@env_name, value)
  defp proper_list?([]), do: true
  defp proper_list?([_head | tail]), do: proper_list?(tail)
  defp proper_list?(_value), do: false
  defp unique?(values), do: length(values) == MapSet.size(MapSet.new(values))

  defp integer_between?(value, minimum, maximum),
    do: is_integer(value) and value >= minimum and value <= maximum

  defp positive(value), do: is_integer(value) and value > 0
  defp nonnegative(value), do: is_integer(value) and value >= 0
  defp exact_keys?(map, keys), do: Map.keys(map) |> Enum.sort() == Enum.sort(keys)
  defp random_id, do: :crypto.strong_rand_bytes(16) |> Base.encode16(case: :lower)
end

File: crates/rekindle-toolchain/src/exec.rs
use crate::{frame, guardian};
use serde_json::{Map, Value, json};
use std::io::Read;
use std::os::unix::fs::PermissionsExt;
use std::process::{Command, Stdio};
use std::sync::mpsc::{self, Receiver, Sender};
use std::thread;
use std::time::Duration;

enum StreamEvent {
    Data(&'static str, Vec<u8>),
    Discarded(&'static str, u64),
    Eof(&'static str),
    Error(String),
}

struct StreamState {
    stdout_sequence: u64,
    stderr_sequence: u64,
    stdout_bytes: u64,
    stderr_bytes: u64,
    discarded_stdout: u64,
    discarded_stderr: u64,
    stdout_eof: bool,
    stderr_eof: bool,
}

struct ExitReport<'a> {
    outcome: &'a str,
    code: Option<i32>,
    signal: Option<i32>,
    cleanup: &'a str,
    stdout_bytes: u64,
    stderr_bytes: u64,
    discarded_stdout: u64,
    discarded_stderr: u64,
}

const MAX_STREAM_BYTES: u64 = 1_048_576;
const MAX_GRACE_MS: u64 = 30_000;
const MIN_KILL_GRACE_MS: u64 = 100;

pub fn run<R: Read>(mut input: R) -> Result<(), String> {
    let spawn = frame::read(&mut input)?.ok_or_else(|| "missing spawn request".to_string())?;
    if !spawn.payload.is_empty() {
        return Err("spawn payload must be empty".into());
    }
    let request = validate_spawn(&spawn.header)?;
    let request_id = request["request_id"].as_str().unwrap().to_owned();

    let mut command = Command::new(request["executable"]["value"].as_str().unwrap());
    command
        .args(
            request["argv"]
                .as_array()
                .unwrap()
                .iter()
                .map(|v| v.as_str().unwrap()),
        )
        .current_dir(request["cwd"].as_str().unwrap())
        .stdin(Stdio::null());
    if request["env_mode"] == "replace" {
        command.env_clear();
    }
    for name in request["env_unset"].as_array().unwrap() {
        command.env_remove(name.as_str().unwrap());
    }
    for pair in request["env_set"].as_array().unwrap() {
        let pair = pair.as_array().unwrap();
        command.env(pair[0].as_str().unwrap(), pair[1].as_str().unwrap());
    }
    let terminate_grace = Duration::from_millis(request["terminate_grace_ms"].as_u64().unwrap());
    let kill_grace = Duration::from_millis(request["kill_grace_ms"].as_u64().unwrap());
    let mut process = match guardian::spawn(command, terminate_grace, kill_grace)? {
        guardian::SpawnResult::Started(process) => process,
        guardian::SpawnResult::Failed => {
            write_exit(
                &request_id,
                &ExitReport {
                    outcome: "spawn_failed",
                    code: None,
                    signal: None,
                    cleanup: "confirmed",
                    stdout_bytes: 0,
                    stderr_bytes: 0,
                    discarded_stdout: 0,
                    discarded_stderr: 0,
                },
            )?;
            return Ok(());
        }
    };
    frame::write(
        &mut std::io::stdout().lock(),
        &json!({
            "v": 1, "type": "started", "request_id": request_id,
            "payload_len": 0, "pid": process.pid, "process_group": process.pid
        }),
        &[],
    )?;

    let (sender, receiver) = mpsc::channel();
    stream_thread(
        process
            .stdout
            .try_clone()
            .map_err(|error| error.to_string())?,
        "stdout",
        sender.clone(),
    );
    stream_thread(
        process
            .stderr
            .try_clone()
            .map_err(|error| error.to_string())?,
        "stderr",
        sender,
    );
    let mut streams = StreamState {
        stdout_sequence: 0,
        stderr_sequence: 0,
        stdout_bytes: 0,
        stderr_bytes: 0,
        discarded_stdout: 0,
        discarded_stderr: 0,
        stdout_eof: false,
        stderr_eof: false,
    };
    let mut exit = None;
    let mut cancelled = false;

    loop {
        drain_streams(&receiver, &request_id, &mut streams)?;
        if exit.is_none() {
            exit = process.try_exit()?;
        }
        if exit.is_some() && streams.stdout_eof && streams.stderr_eof {
            break;
        }

        if !cancelled && stdin_ready()? {
            match frame::read(&mut input)? {
                Some(cancel) => {
                    validate_cancel(&cancel, &request_id)?;
                    process.cancel()?;
                    cancelled = true;
                }
                None => {
                    process.cancel()?;
                    cancelled = true;
                }
            }
        }
        thread::sleep(Duration::from_millis(5));
    }

    let exit = exit.ok_or_else(|| "guardian terminal report missing".to_string())?;
    write_exit(
        &request_id,
        &ExitReport {
            outcome: exit.outcome,
            code: exit.code,
            signal: exit.signal,
            cleanup: exit.cleanup,
            stdout_bytes: streams.stdout_bytes,
            stderr_bytes: streams.stderr_bytes,
            discarded_stdout: streams.discarded_stdout,
            discarded_stderr: streams.discarded_stderr,
        },
    )
}

fn validate_spawn(value: &Value) -> Result<&Map<String, Value>, String> {
    const KEYS: &[&str] = &[
        "v",
        "type",
        "request_id",
        "payload_len",
        "executable",
        "argv",
        "cwd",
        "env_mode",
        "env_set",
        "env_unset",
        "terminate_grace_ms",
        "kill_grace_ms",
    ];
    if !frame::exact_keys(value, KEYS) {
        return Err("invalid spawn fields".into());
    }
    let object = value.as_object().unwrap();
    let executable = object["executable"]
        .as_object()
        .ok_or("invalid executable")?;
    let executable_keys = executable.keys().map(String::as_str).collect::<Vec<_>>();
    let executable_path = executable
        .get("value")
        .and_then(Value::as_str)
        .unwrap_or_default();
    if object["type"] != "spawn"
        || object["payload_len"] != 0
        || executable_keys.len() != 2
        || executable.get("kind") != Some(&Value::String("path".into()))
        || !absolute(executable_path)
        || !absolute(object["cwd"].as_str().unwrap_or_default())
        || !qualified_executable(executable_path)
        || !qualified_directory(object["cwd"].as_str().unwrap_or_default())
        || !strings(&object["argv"])
        || !matches!(object["env_mode"].as_str(), Some("inherit" | "replace"))
        || !environment(&object["env_set"], &object["env_unset"])
        || !milliseconds_in(&object["terminate_grace_ms"], 0, MAX_GRACE_MS)
        || !milliseconds_in(&object["kill_grace_ms"], MIN_KILL_GRACE_MS, MAX_GRACE_MS)
    {
        return Err("invalid spawn request".into());
    }
    Ok(object)
}

fn validate_cancel(frame_value: &frame::Frame, request_id: &str) -> Result<(), String> {
    const KEYS: &[&str] = &["v", "type", "request_id", "payload_len", "reason"];
    let value = &frame_value.header;
    if !frame_value.payload.is_empty()
        || !frame::exact_keys(value, KEYS)
        || value["type"] != "cancel"
        || value["request_id"] != request_id
        || value["payload_len"] != 0
        || !matches!(
            value["reason"].as_str(),
            Some("obsolete" | "timeout" | "shutdown" | "caller")
        )
    {
        return Err("invalid cancel".into());
    }
    Ok(())
}

fn stream_thread<R: Read + Send + 'static>(
    mut reader: R,
    kind: &'static str,
    sender: Sender<StreamEvent>,
) {
    thread::spawn(move || {
        let mut buffer = vec![0_u8; 64 * 1024];
        let mut emitted = 0_u64;
        loop {
            match reader.read(&mut buffer) {
                Ok(0) => {
                    let _ = sender.send(StreamEvent::Eof(kind));
                    break;
                }
                Ok(size) => {
                    let available = MAX_STREAM_BYTES.saturating_sub(emitted) as usize;
                    let retained = size.min(available);
                    if retained > 0 {
                        emitted += retained as u64;
                        if sender
                            .send(StreamEvent::Data(kind, buffer[..retained].to_vec()))
                            .is_err()
                        {
                            break;
                        }
                    }
                    if retained < size
                        && sender
                            .send(StreamEvent::Discarded(kind, (size - retained) as u64))
                            .is_err()
                    {
                        break;
                    }
                }
                Err(error) => {
                    let _ = sender.send(StreamEvent::Error(error.to_string()));
                    break;
                }
            }
        }
    });
}

fn drain_streams(
    receiver: &Receiver<StreamEvent>,
    request_id: &str,
    state: &mut StreamState,
) -> Result<(), String> {
    while let Ok(event) = receiver.try_recv() {
        match event {
            StreamEvent::Data(kind, bytes) => write_stream(request_id, kind, false, bytes, state)?,
            StreamEvent::Discarded("stdout", bytes) => state.discarded_stdout += bytes,
            StreamEvent::Discarded("stderr", bytes) => state.discarded_stderr += bytes,
            StreamEvent::Discarded(_, _) => return Err("invalid stream kind".into()),
            StreamEvent::Eof(kind) => write_stream(request_id, kind, true, Vec::new(), state)?,
            StreamEvent::Error(error) => return Err(format!("stream read failed: {error}")),
        }
    }
    Ok(())
}

fn write_stream(
    request_id: &str,
    kind: &str,
    eof: bool,
    bytes: Vec<u8>,
    state: &mut StreamState,
) -> Result<(), String> {
    let (sequence, total, ended) = if kind == "stdout" {
        (
            &mut state.stdout_sequence,
            &mut state.stdout_bytes,
            &mut state.stdout_eof,
        )
    } else {
        (
            &mut state.stderr_sequence,
            &mut state.stderr_bytes,
            &mut state.stderr_eof,
        )
    };
    *total += bytes.len() as u64;
    frame::write(
        &mut std::io::stdout().lock(),
        &json!({
            "v": 1, "type": kind, "request_id": request_id,
            "payload_len": bytes.len(), "sequence": *sequence, "eof": eof
        }),
        &bytes,
    )?;
    *sequence += 1;
    *ended = eof;
    Ok(())
}

fn stdin_ready() -> Result<bool, String> {
    let mut descriptor = libc::pollfd {
        fd: 0,
        events: libc::POLLIN | libc::POLLHUP,
        revents: 0,
    };
    let result = unsafe { libc::poll(&mut descriptor, 1, 0) };
    if result < 0 {
        Err(std::io::Error::last_os_error().to_string())
    } else {
        Ok(result > 0 && descriptor.revents != 0)
    }
}

fn write_exit(request_id: &str, report: &ExitReport<'_>) -> Result<(), String> {
    frame::write(
        &mut std::io::stdout().lock(),
        &json!({
            "v": 1, "type": "exit", "request_id": request_id, "payload_len": 0,
            "outcome": report.outcome, "code": report.code, "signal": report.signal,
            "cleanup": report.cleanup, "stdout_bytes": report.stdout_bytes,
            "stderr_bytes": report.stderr_bytes,
            "discarded_stdout": report.discarded_stdout,
            "discarded_stderr": report.discarded_stderr
        }),
        &[],
    )
}

fn absolute(value: &str) -> bool {
    value.starts_with('/') && !value.as_bytes().contains(&0)
}

fn strings(value: &Value) -> bool {
    value.as_array().is_some_and(|values| {
        values
            .iter()
            .all(|v| v.as_str().is_some_and(|s| !s.as_bytes().contains(&0)))
    })
}

fn environment(set: &Value, unset: &Value) -> bool {
    let Some(set) = set.as_array() else {
        return false;
    };
    let Some(unset) = unset.as_array() else {
        return false;
    };
    let mut names = std::collections::BTreeSet::new();
    for pair in set {
        let Some(pair) = pair.as_array() else {
            return false;
        };
        if pair.len() != 2
            || !valid_env(pair[0].as_str())
            || pair[1].as_str().is_none()
            || !names.insert(pair[0].as_str().unwrap())
        {
            return false;
        }
    }
    let mut unset_names = std::collections::BTreeSet::new();
    unset.iter().all(|name| {
        valid_env(name.as_str())
            && !names.contains(name.as_str().unwrap())
            && unset_names.insert(name.as_str().unwrap())
    })
}

fn qualified_executable(path: &str) -> bool {
    let Ok(metadata) = std::fs::symlink_metadata(path) else {
        return false;
    };
    metadata.file_type().is_file()
        && metadata.permissions().mode() & 0o100 != 0
        && no_symlink_components(path)
}

fn qualified_directory(path: &str) -> bool {
    std::fs::symlink_metadata(path)
        .is_ok_and(|metadata| metadata.file_type().is_dir() && no_symlink_components(path))
}

fn no_symlink_components(path: &str) -> bool {
    let mut current = std::path::PathBuf::new();
    for component in std::path::Path::new(path).components() {
        current.push(component.as_os_str());
        if matches!(component, std::path::Component::RootDir) {
            continue;
        }
        if !std::fs::symlink_metadata(&current)
            .is_ok_and(|metadata| !metadata.file_type().is_symlink())
        {
            return false;
        }
    }
    true
}

fn valid_env(value: Option<&str>) -> bool {
    value.is_some_and(|value| {
        let mut chars = value.chars();
        chars
            .next()
            .is_some_and(|c| c == '_' || c.is_ascii_alphabetic())
            && chars.all(|c| c == '_' || c.is_ascii_alphanumeric())
            && value.len() <= 128
    })
}

fn milliseconds_in(value: &Value, minimum: u64, maximum: u64) -> bool {
    value
        .as_u64()
        .is_some_and(|milliseconds| (minimum..=maximum).contains(&milliseconds))
}

#[cfg(test)]
mod tests {
    use super::validate_spawn;
    use serde_json::{Value, json};

    #[test]
    fn admits_only_the_exact_grace_boundaries() {
        for (terminate, kill) in [(0, 100), (30_000, 30_000)] {
            assert!(validate_spawn(&spawn(terminate.into(), kill.into())).is_ok());
        }

        for (terminate, kill) in [(-1, 100), (30_001, 100), (0, 99), (0, 30_001)] {
            assert!(validate_spawn(&spawn(terminate.into(), kill.into())).is_err());
        }

        for (terminate, kill) in [
            (json!(0.0), json!(100)),
            (json!(0), json!(100.0)),
            (json!("0"), json!(100)),
            (json!(0), Value::Null),
        ] {
            assert!(validate_spawn(&spawn(terminate, kill)).is_err());
        }
    }

    fn spawn(terminate: Value, kill: Value) -> Value {
        let executable = std::env::current_exe().expect("test executable");

        json!({
            "v": 1,
            "type": "spawn",
            "request_id": "0123456789abcdef0123456789abcdef",
            "payload_len": 0,
            "executable": {"kind": "path", "value": executable},
            "argv": [],
            "cwd": "/",
            "env_mode": "replace",
            "env_set": [],
            "env_unset": [],
            "terminate_grace_ms": terminate,
            "kill_grace_ms": kill
        })
    }
}
